### PR TITLE
add scan edge interface

### PR DESCRIPTION
**Pull request by jievince**
Depends on https://github.com/vesoft-inc/nebula-common/pull/516

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -34,3 +34,7 @@
 _build
 build
 modules
+
+.vscode
+
+examples/
```
Comment (Shylock-Hg) [inline .gitignore]: Why ignore examples?

File: include/nebula/sclient/StorageClient.h [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    int32_t getSpaceVidLen(std::string spaceName);
+
+    std::vector<int32_t> getParts(std::string spaceName);
+
+    std::unordered_map<int32_t, std::vector<nebula::MetaHostAddr>> getPartsAlloc(
+        std::string spaceName);
+
+    std::unordered_map<int32_t, nebula::MetaHostAddr> getPartsLeader(std::string spaceName);
+
+    ScanEdgeIter scanEdgeWithPart(std::string spaceName,
```
Comment (laura-ding) [inline include/nebula/sclient/StorageClient.h:69]: why only support the scan with the part?

File: src/sclient/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2018 vesoft inc. All rights reserved.
```
Comment (laura-ding) [inline src/sclient/CMakeLists.txt:1]: all the year set to 2021

File: examples/StorageClientExample.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                                             10,
+                                             0,
+                                             std::numeric_limits<int64_t>::max(),
+                                             "",
+                                             true,
+                                             true);
+    std::cout << "scan edge..." << std::endl;
+    while (scanEdgeIter.hasNext()) {
+        std::cout << "-------------------------" << std::endl;
+        nebula::DataSet expected({"likeness"});
```
Comment (laura-ding) [inline examples/StorageClientExample.cpp]: not used

File: examples/StorageClientExample.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                                          0,
+                                          std::numeric_limits<int64_t>::max(),
+                                          "",
+                                          true,
+                                          true);
+
+    std::cout << "scan vertex" << std::endl;
+    while (scanVertexIter.hasNext()) {
+        std::cout << "-------------------------" << std::endl;
+        nebula::DataSet expected({"likeness"});
```
Comment (laura-ding) [inline examples/StorageClientExample.cpp]: ditto

File: examples/StorageClientExample.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+#include <nebula/sclient/Init.h>
+#include <nebula/sclient/StorageClient.h>
+
+int main(int argc, char* argv[]) {
+    nebula::init(&argc, &argv);
+
+    nebula::StorageClient c({nebula::MetaHostAddr("localhost", 45996)});
+
+    auto scanEdgeIter = c.scanEdgeWithPart("nba",
+                                             1,
```
Comment (laura-ding) [inline examples/StorageClientExample.cpp]: format

File: examples/StorageClientExample.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    while (scanEdgeIter.hasNext()) {
+        std::cout << "-------------------------" << std::endl;
+        nebula::DataSet expected({"likeness"});
+        nebula::DataSet ds = scanEdgeIter.next();
+        std::cout << ds << std::endl;
+        std::cout << "+++++++++++++++++++++++++" << std::endl;
+    }
+
+    auto scanVertexIter = c.scanVertexWithPart("nba",
+                                          1,
```
Comment (laura-ding) [inline examples/StorageClientExample.cpp]: ditto

File: include/nebula/sclient/StorageClient.h [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    ScanVertexIter scanVertexWithPart(std::string spaceName,
+                                      int32_t partId,
+                                      std::string tagName,
+                                      std::vector<std::string> propNames,
+                                      int64_t limit,
+                                      int64_t startTime,
+                                      int64_t endTime,
+                                      std::string filter,
+                                      bool onlyLatestVersion,
+                                      bool enableReadFromFollower);
```
Comment (laura-ding) [inline include/nebula/sclient/StorageClient.h]: set default value

File: src/sclient/tests/StorageClientTest.cpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            std::cout << "partReplicaSize[" << p.first << "]=" << p.second.size() << std::endl;
+        }
+        std::cout << "-----------------------" << std::endl;
+    }
+
+    static void runGetPartsLeader(nebula::StorageClient &c) {
+        std::cout << "get parts leader of space" << std::endl;
+        auto partsLeader = c.getPartsLeader("nba");
+        for (auto p : partsLeader) {
+            std::cout << "partLeader[" << p.first << "]=" << p.second << std::endl;
```
Comment (laura-ding) [inline src/sclient/tests/StorageClientTest.cpp]: if you want to test the result, no only print, please check the result

File: src/sclient/tests/StorageClientTest.cpp [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                                                 true);
+        {
+            EXPECT_EQ(scanResultIter.hasNext(), true);
+            nebula::DataSet ds = scanResultIter.next();
+            EXPECT_EQ(ds, nebula::DataSet());
+            EXPECT_EQ(scanResultIter.hasNext_, false);
+            EXPECT_EQ(scanResultIter.nextCursor_, "");
+        }
+
+        scanResultIter = c.scanEdgeWithPart("nba",
```
Comment (laura-ding) [inline src/sclient/tests/StorageClientTest.cpp:69]: Your test data is not inserted, so the test here is no used. Please perfect the test

File: src/sclient/StorageClient.cpp [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+    }
+    DCHECK(result.size() == partsAlloc.size());
+
+    return result;
+}
+
+ScanEdgeIter StorageClient::scanEdgeWithPart(std::string spaceName,
+                                             int32_t partId,
+                                             std::string edgeName,
+                                             std::vector<std::string> propNames,
```
Comment (laura-ding) [inline src/sclient/StorageClient.cpp:162]: the keys of the edge is unknown by users, you need to add the key prop of the edge.

```
'_src'
'_type'
'_rank'
'_dst'
```

File: src/sclient/StorageClient.cpp [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+    req->set_only_latest_version(onlyLatestVersion);
+    req->set_enable_read_from_follower(enableReadFromFollower);
+
+    return ScanEdgeIter(sclient_.get(), req);
+}
+
+ScanVertexIter StorageClient::scanVertexWithPart(std::string spaceName,
+                                                 int32_t partId,
+                                                 std::string tagName,
+                                                 std::vector<std::string> propNames,
```
Comment (laura-ding) [inline src/sclient/StorageClient.cpp:204]: the keys of the vertex is unknown by users, you need to add the key prop of the edge.

```
'_vid'
```

File: src/sclient/ScanVertexIter.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    return *this;
+}
+
+ScanVertexIter::~ScanVertexIter() {
+    delete req_;
+}
+
+DataSet ScanVertexIter::next() {
+    if (!hasNext_) {
+        LOG(ERROR) << "hasNext() == false !";
```
Comment (laura-ding) [inline src/sclient/ScanVertexIter.cpp:55]: delete the log

File: src/sclient/ScanVertexIter.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        return DataSet();
+    }
+    DCHECK(!!req_);
+    req_->set_cursor(std::move(nextCursor_));
```
Comment (laura-ding) [inline src/sclient/ScanVertexIter.cpp:59]: why use move? it will call next multi time

File: src/sclient/ScanVertexIter.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        this->hasNext_ = false;
+        return DataSet();
+    }
+    this->hasNext_ = scanResponse.get_has_next();
+    std::string* cursor = scanResponse.get_next_cursor();
+    if (cursor) {
+        this->nextCursor_ = *cursor;
+    }
+    auto data = scanResponse.get_vertex_data();
+    return data;
```
Comment (laura-ding) [inline src/sclient/ScanVertexIter.cpp]: return scanResponse.get_vertex_data();

File: src/sclient/ScanEdgeIter.cpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        this->hasNext_ = false;
+        return DataSet();
+    }
+    this->hasNext_ = scanResponse.get_has_next();
+    std::string* cursor = scanResponse.get_next_cursor();
+    if (cursor) {
+        this->nextCursor_ = *cursor;
+    }
+    auto data = scanResponse.get_edge_data();
+    return data;
```
Comment (laura-ding) [inline src/sclient/ScanEdgeIter.cpp]: return scanResponse.get_vertex_data();

File: include/nebula/sclient/StorageClient.h (hunk 2 above)
Comment (jievince) [reply include/nebula/sclient/StorageClient.h:69]: I just need this interface currently

File: src/sclient/CMakeLists.txt (hunk 3 above)
Comment (jievince) [reply src/sclient/CMakeLists.txt:1]: ok

File: include/nebula/sclient/StorageClient.h (hunk 8 above)
Comment (jievince) [reply include/nebula/sclient/StorageClient.h]: ok, thanks

File: src/sclient/StorageClient.cpp (hunk 11 above)
Comment (jievince) [reply src/sclient/StorageClient.cpp:162]: currently I just need the "_src" and "_dst"

File: src/sclient/ScanVertexIter.cpp (hunk 14 above)
Comment (jievince) [reply src/sclient/ScanVertexIter.cpp:59]: because the value of nextCursor_ will change next time

File: src/sclient/ScanEdgeIter.cpp (hunk 16 above)
Comment (jievince) [reply src/sclient/ScanEdgeIter.cpp]: Ok, thanks

File: src/sclient/ScanVertexIter.cpp (hunk 15 above)
Comment (jievince) [reply src/sclient/ScanVertexIter.cpp]: ok

File: src/sclient/ScanVertexIter.cpp (hunk 13 above)
Comment (jievince) [reply src/sclient/ScanVertexIter.cpp:55]: Why delete it

File: .gitignore (hunk 1 above)
Comment (jievince) [reply .gitignore]: has removed.

**Review by nevermore3 — DISMISSED** [review]

**Pull request closed without merging**